### DISCUSSION TITLE
Fixing name collisions in C

### DISCUSSION THREAD

**Discussion by EmilyBourne** [pyccel/pyccel#1149] [Ideas]
**Describe the bug**
C has no way to rename functions, nor does it have namespaces. This makes collisions much more likely when using more than one file. Furthermore there is no way to import part of a file. This makes collisions even more likely.

To fix this problem I suggest pre-pending the name of the module to each function/variable/class defined within that module in the C code. I.e:
`f` -> `mod__f`

This should prevent almost all collisions. Collisions are still possible if the user names his variables using the following pattern:

NAME1__NAME2.py containing:
    - function : `f()`
NAME1.py containing:
    - function `NAME2__f()`

But this seems unlikely, especially due to the double `_`

**Test Code**
Test files can be found [here](https://github.com/pyccel/pyccel/tree/master/tests/pyccel/scripts/import_syntax). When working they can be activated for C in [this file](https://github.com/pyccel/pyccel/blob/master/tests/pyccel/test_pyccel.py)

**Language**
C

**Additional Information**
I suspect the difficulty here will be at the function call. The namespace must be used to find the Import which created the object in order to find the module it has come from and it's original name

**Comment by saidctb** [comment]
![image](https://user-images.githubusercontent.com/11571079/130569203-0ca156eb-2906-4011-8a76-8f090fdcb7a6.png)
how about doing something like this?

**Comment by EmilyBourne** [comment]
@saidctb This solution is suggested on stack overflow [here](https://stackoverflow.com/a/28535585/5957195), however the comments on this solution indicate that it adds a lot of overhead

**Comment by EmilyBourne** [comment]
Plus I am not sure that there is a huge readability difference between:
```
yourApi.doFoo(42)
```
and
```
yourApi__doFoo(42)
```

**Comment by saidctb** [comment]
I see, yes there is an overhead

**Comment by yguclu** [comment]
How much overhead? Unless we want to get those functions inlined, I suspect it's minimal.

**Comment by EmilyBourne** [comment]
> How much overhead? Unless we want to get those functions inlined, I suspect it's minimal.

I have not tested, so I'm not sure. But a discussion [here](https://stackoverflow.com/a/2438558/5957195) claims that:
> using function pointers may prevent compiler optimizations (inlining) and processor optimizations (branch prediction)

**Comment by EmilyBourne** [comment]
It is possible that the overhead is not huge here, but given that our aim is to be as fast as possible and that we don't need this I don't see why we should add that overhead. Do you both find:
```
yourApi.doFoo(42)
```
significantly more readable than:
```
yourApi__doFoo(42)
```
?
Or are you worried about the unlikely name collisions? Or are you thinking about the readability of files which do not import other files?

**Comment by yguclu** [comment]
I just wanted us to know the order of magnitude of the overhead in order to make an informed decision.

In a large project two Python modules with the same name could exist in different sub-directories, hence we will need to prepend the namespace with the directory name. This will make the code very hard to read, so we may want to perform name mangling only if the user asks for it explicitly.

**Comment by saidctb** [comment]
This is the timing that I got comparing between a macro and a function call in a loop of `1000000` iterations
![image](https://user-images.githubusercontent.com/11571079/131120039-ae73bf52-e298-4d8a-b333-9a2e130782ea.png)

```c
#include <stdio.h>
#include <time.h>
#include "f1.h"

#define add_macro(a, b) ((a)+(b))


void show_time(char *str, struct timespec *tv1, struct timespec *tv2)
{
  printf("%s: %ld seconds\n", str, tv2->tv_sec-tv1->tv_sec);
  printf("%s: %ld nano seconds\n", str, tv2->tv_nsec-tv1->tv_nsec);
}

int main(void)
{
  struct timespec  tv1, tv2;
  unsigned long long int i, N, val=0;

  printf("Enter number of iterations: ");
  int err = scanf("%lld", &N);

  timespec_get(&tv1, TIME_UTC);
  for(i = 0;i < N;++i)
    val = api1.add(val, 1);
  timespec_get(&tv2, TIME_UTC);

  show_time("function call timing:", &tv1, &tv2);
  printf("sum value: %lld\n\n", val);

  val = 0;
  timespec_get(&tv1, TIME_UTC);
  for(i = 0;i < N;++i)
    val = add_macro(val, 1);
  timespec_get(&tv2, TIME_UTC);

  show_time("macro timing", &tv1, &tv2);
  printf("sum value: %lld\n", val);
  return 0;
}
```

**Comment by EmilyBourne** [comment]
@saidctb Can you compare a function call and a function call via a pointer? This is closer to the difference that we are talking about.

As a second point, I wonder how this would affect the cwrapper given that it uses the c printer.

**Comment by saidctb** [comment]
@EmilyBourne what do you mean?, in a c program generated by pyccel we don't have any of that 
but if we use the cwrapper, the overhead will be negligible compared to the python function call

**Comment by EmilyBourne** [comment]
@saidctb Are you just asking about my second point here?

I was wondering if there were any changes that would have to be made to the cwrapper. I think you are right and that it shouldn't change anything for the names of the functions in the cwrapper. I was thinking about the calls to fortran functions though. Although I suppose the naming for these functions will need to be handled differently anyway (and indeed already is), so maybe this is a non-issue

**Comment by saidctb** [comment]
> @saidctb Are you just asking about my second point here?
> 
> I was wondering if there were any changes that would have to be made to the cwrapper. I think you are right and that it shouldn't change anything for the names of the functions in the cwrapper. I was thinking about the calls to fortran functions though. Although I suppose the naming for these functions will need to be handled differently anyway (and indeed already is), so maybe this is a non-issue

I'm talking about the first point

**Comment by EmilyBourne** [comment]
@saidctb 
> I'm talking about the first point

In your example you compared a macro and a function pointer. Currently we are using basic functions. Macros are always going to be faster (but slower to compile).

So far we have suggested 2 solutions to the collision problem:
1. Rename (uses functions)
2. Function pointers

Neither of these are macros so I don't see the relevance in comparing macros and function pointers. I would be more interested in the results of the same experiment but with a function and a function pointer:

```
 timespec_get(&tv1, TIME_UTC);
  for(i = 0;i < N;++i)
    val = api1.add(val, 1);
  timespec_get(&tv2, TIME_UTC);

  show_time("function call timing:", &tv1, &tv2);
  printf("sum value: %lld\n\n", val);

  val = 0;
  timespec_get(&tv1, TIME_UTC);
  for(i = 0;i < N;++i)
    val = add(val, 1);
  timespec_get(&tv2, TIME_UTC);
```

**Comment by saidctb** [comment]
this is what we get when comparing it with a simple function call
![image](https://user-images.githubusercontent.com/11571079/131142677-6e74c20e-d8cd-46fd-94f9-58ad244eed65.png)

**Comment by EmilyBourne** [comment]
So then the useful comparison is:
```
function call timing         : 6306546
function call pointer timing : 5915644
```
So it seems there is no noticeable loss here then

**Comment by saidctb** [comment]
ah no I made a mistake , I should have changed the print message.
this is the difference
```
function call pointer timing   : 6306546
function call timing           : 275734
```

**Comment by yguclu** [comment]
So how much time per call?

**Comment by saidctb** [comment]
@yguclu 
it's
```
function call pointer timing   : 6.306546 nano seconds per call
function call timing           : 0.275734 nano seconds per call
```

**Comment by EmilyBourne** [comment]
I just noticed that this discussion never reached a decision. A factor of 30 seems like a lot to me. So I still think that we ought to fix this with a naming convention rather than a structure containing function pointers

**Comment by EmilyBourne** [comment]
While investigating bugs on windows with the updated compiler I realised there is another problem with the function pointer solution.

C does not usually mangle names. As a result, even if objects are only declared in .c file, they remain in the global scope. Therefore when linking certain compilers will give errors complaining about multiple definitions of the same object.

For example this happened on windows for the variable `initialised` used by modules in the initialisation, despite this variable not being mentioned in the .h file. The error was:
```
c:/programdata/chocolatey/lib/mingw/tools/install/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/11.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: 
D:\a\pyccel\pyccel\tests\pyccel\scripts\__pyccel__\module_init.o:module_init.c:(.bss+0x0): multiple definition of `initialised'; 
D:\a\pyccel\pyccel\tests\pyccel\scripts\__pyccel__\runtest_module_init.o:runtest_module_init.c:(.bss+0x0): first defined here
```

The solution therefore seems clear. I have changed this to a discussion and will open a fresh issue so it is clear what needs to be done for anyone seeing this issue for the first time.

**Comment by EmilyBourne** [comment]
Should the same method be used when creating class methods?
